modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionDataStorage.java

```diff
@@ -0,0 +1,136 @@
+     *
+     * @return Future that's completed when flushing of the data is completed.
+     */
+    CompletableFuture<Void> flush();
+
+    /**
+     * Index of the highest write command applied to the storage. {@code 0} if index is unknown.
+     */
+    long lastAppliedIndex();
+
+    /**
+     * Sets the last applied index value.
+     */
+    void lastAppliedIndex(long lastAppliedIndex) throws StorageException;
+
+    /**
+     * Creates (or replaces) an uncommitted (aka pending) version, assigned to the given transaction id.
+     * In details:
+     * - if there is no uncommitted version, a new uncommitted version is added
+     * - if there is an uncommitted version belonging to the same transaction, it gets replaced by the given version
+     * - if there is an uncommitted version belonging to a different transaction, {@link TxIdMismatchException} is thrown
+     *
+     * <p>This must be called under a lock acquired using {@link #acquirePartitionSnapshotsReadLock()}.
+     *
+     * @param rowId Row id.
+     * @param row Binary row to update. Key only row means value removal.
+     * @param txId Transaction id.
+     * @param commitTableId Commit table id.
+     * @param commitPartitionId Commit partitionId.
+     * @return Previous uncommitted row version associated with the row id, or {@code null} if no uncommitted version
+     *     exists before this call
+     * @throws TxIdMismatchException If there's another pending update associated with different transaction id.
+     * @throws StorageException If failed to write data to the storage.
+     */
+    @Nullable BinaryRow addWrite(RowId rowId, @Nullable BinaryRow row, UUID txId, UUID commitTableId, int commitPartitionId)
+            throws TxIdMismatchException, StorageException;
+
+    /**
+     * Aborts a pending update of the ongoing uncommitted transaction. Invoked during rollback.
+     *
+     * <p>This must be called under a lock acquired using {@link #acquirePartitionSnapshotsReadLock()}.
+     *
+     * @param rowId Row id.
+     * @return Previous uncommitted row version associated with the row id.
+     * @throws StorageException If failed to write data to the storage.
+     */
+    @Nullable BinaryRow abortWrite(RowId rowId) throws StorageException;
+
+    /**
+     * Commits a pending update of the ongoing transaction. Invoked during commit. Committed value will be versioned by the given timestamp.
+     *
+     * <p>This must be called under a lock acquired using {@link #acquirePartitionSnapshotsReadLock()}.
+     *
+     * @param rowId Row id.
+     * @param timestamp Timestamp to associate with committed value.
+     * @throws StorageException If failed to write data to the storage.
+     */
+    void commitWrite(RowId rowId, HybridTimestamp timestamp) throws StorageException;
+
+    /**
+     * Returns the underlying {@link MvPartitionStorage}. Only for tests!
+     *
+     * @return Underlying {@link MvPartitionStorage}.
+     */
+    @TestOnly
```

Comment:
Strictly speaking such a method breaks encapsulation; if we have it then there is some problem with our interface. For tests it might be ok, but in production code I would try to avoid such methods.

Reply:
It doesn't break encapsulation, we can rely on the fact that each partition owns mv storage, it's not a secret, it's a part of the design

Reply:
1. Now we don't need it in the production code
2. If we had, we'd violate 'tell, not ask' principle. Anything that needs to be done on `MvPartitionStorage` by a client of `PartitionDataStorage` could be done via its interface, this makes clients simpler and makes writing unit tests easier